File: training-data/perl/skip_list.pl
use strict;
use warnings;

package SkipList;

sub new {
    my ($class, $max_level) = @_;
    my $self = {
        max_level => $max_level,
        head      => { value => undef, forward => [] },
    };
    push @{$self->{head}{forward}}, undef for 1 .. $max_level;
    return bless $self, $class;
}

sub _random_level {
    my $self = shift;
    my $level = 1;
    while ($level < $self->{max_level} && rand() < 0.5) {
        $level++;
    }
    return $level;
}

sub insert {
    my ($self, $value) = @_;
    my $node = $self->{head};
    for (my $lvl = $self->{max_level} - 1; $lvl >= 0; $lvl--) {
        while ($node->{forward}[$lvl] && $node->{forward}[$lvl]{value} < $value) {
            $node = $node->{forward}[$lvl];
        }
    }
    my $level = $self->_random_level();
    my $new_node = { value => $value, forward => [(undef) x $level] };
    $node = $self->{head};
    for (my $lvl = $self->{max_level} - 1; $lvl >= 0; $lvl--) {
        while ($node->{forward}[$lvl] && $node->{forward}[$lvl]{value} < $value) {
            $node = $node->{forward}[$lvl];
        }
        if ($lvl < $level) {
            $new_node->{forward}[$lvl] = $node->{forward}[$lvl];
            $node->{forward}[$lvl] = $new_node;
        }
    }
}

sub contains {
    my ($self, $value) = @_;
    my $node = $self->{head};
    for (my $lvl = $self->{max_level} - 1; $lvl >= 0; $lvl--) {
        while ($node->{forward}[$lvl] && $node->{forward}[$lvl]{value} < $value) {
            $node = $node->{forward}[$lvl];
        }
    }
    $node = $node->{forward}[0];
    return $node && $node->{value} == $value;
}

sub to_array {
    my $self = shift;
    my @result;
    my $node = $self->{head}{forward}[0];
    while ($node) {
        push @result, $node->{value};
        $node = $node->{forward}[0];
    }
    return @result;
}

package main;

srand(42);
my $list = SkipList->new(4);
$list->insert($_) for (3, 6, 7, 9, 12, 19, 17);
print join(" ", $list->to_array()), "\n";
print "contains 9: ", ($list->contains(9) ? "yes" : "no"), "\n";
print "contains 10: ", ($list->contains(10) ? "yes" : "no"), "\n";

File: training-data/go/skip_list.go
package main

import (
	"fmt"
	"math/rand"
)

const maxLevel = 4

type skipNode struct {
	value   int
	forward []*skipNode
}

type SkipList struct {
	head  *skipNode
	level int
	rng   *rand.Rand
}

func NewSkipList() *SkipList {
	return &SkipList{
		head:  &skipNode{forward: make([]*skipNode, maxLevel)},
		level: 1,
		rng:   rand.New(rand.NewSource(1)),
	}
}

func (s *SkipList) randomLevel() int {
	level := 1
	for level < maxLevel && s.rng.Intn(2) == 0 {
		level++
	}
	return level
}

func (s *SkipList) Insert(value int) {
	update := make([]*skipNode, maxLevel)
	current := s.head
	for i := s.level - 1; i >= 0; i-- {
		for current.forward[i] != nil && current.forward[i].value < value {
			current = current.forward[i]
		}
		update[i] = current
	}

	level := s.randomLevel()
	if level > s.level {
		for i := s.level; i < level; i++ {
			update[i] = s.head
		}
		s.level = level
	}

	node := &skipNode{value: value, forward: make([]*skipNode, level)}
	for i := 0; i < level; i++ {
		node.forward[i] = update[i].forward[i]
		update[i].forward[i] = node
	}
}

func (s *SkipList) Contains(value int) bool {
	current := s.head
	for i := s.level - 1; i >= 0; i-- {
		for current.forward[i] != nil && current.forward[i].value < value {
			current = current.forward[i]
		}
	}
	current = current.forward[0]
	return current != nil && current.value == value
}

func (s *SkipList) ToSlice() []int {
	var result []int
	for node := s.head.forward[0]; node != nil; node = node.forward[0] {
		result = append(result, node.value)
	}
	return result
}

func main() {
	list := NewSkipList()
	for _, v := range []int{3, 6, 7, 9, 12, 19, 17} {
		list.Insert(v)
	}

	fmt.Println(list.ToSlice())
	fmt.Println(list.Contains(9), list.Contains(100))
}
